src/main/java/com/amazon/ion/impl/BufferedOutputStreamFastAppendable.kt

```diff
@@ -0,0 +1,191 @@
+
+    override fun append(csq: CharSequence, start: Int, end: Int): Appendable {
+        for (i in start until end) {
+            append(csq[i])
+        }
+        return this
+    }
+
+    override fun appendAscii(c: Char) {
+        assert(c.code < 0x80)
+        write(c.code)
+    }
+
+    override fun appendAscii(csq: CharSequence) = appendAscii(csq, 0, csq.length)
+
+    override fun appendAscii(csq: CharSequence, start: Int, end: Int) {
+        if (csq is String) {
+            // Using deprecated String.getBytes intentionally, since it is
+            // correct behavior in this case, and much faster.
+            var pos = start
+            val len = end - start
+            if (len > current.remaining() && current.remaining() < maxBlockWaste && len < allocator.blockSize) {
+                current = nextBlock()
+            }
+            while (true) {
+                val copyAmount = minOf(current.remaining(), end - pos)
+                csq.copyAsciiBytes(pos, pos + copyAmount, current.data, current.limit)
+                current.limit += copyAmount
+                pos += copyAmount
+                if (pos >= end) return
+                current = nextBlock()
+            }
+        } else {
+            append(csq, start, end)
+        }
+    }
+
+    override fun appendUtf16(c: Char) {
+        assert(c.code >= 0x80)
+        if (current.remaining() < 3) {
+            current = nextBlock()
+        }
+        if (c.code < 0x800) {
+            current.data[current.limit++] = (0xff and (0xC0 or (c.code shr 6))).toByte()
+            current.data[current.limit++] = (0xff and (0x80 or (c.code and 0x3F))).toByte()
+        } else if (c.code < 0x10000) {
+            current.data[current.limit++] = (0xff and (0xE0 or (c.code shr 12))).toByte()
+            current.data[current.limit++] = (0xff and (0x80 or (c.code shr 6 and 0x3F))).toByte()
+            current.data[current.limit++] = (0xff and (0x80 or (c.code and 0x3F))).toByte()
+        }
+    }
+
+    override fun appendUtf16Surrogate(leadSurrogate: Char, trailSurrogate: Char) {
+        // Here we must convert a UTF-16 surrogate pair to UTF-8 bytes.
+        val c = _Private_IonConstants.makeUnicodeScalar(leadSurrogate.code, trailSurrogate.code)
+        assert(c >= 0x10000)
+        if (current.remaining() < 4) {
+            current = nextBlock()
+        }
+        current.data[current.limit++] = (0xff and (0xF0 or (c shr 18))).toByte()
+        current.data[current.limit++] = (0xff and (0x80 or (c shr 12 and 0x3F))).toByte()
+        current.data[current.limit++] = (0xff and (0x80 or (c shr 6 and 0x3F))).toByte()
+        current.data[current.limit++] = (0xff and (0x80 or (c and 0x3F))).toByte()
+    }
+
+    /** Helper function to wrap [java.lang.String.getBytes]. */
+    private fun String.copyAsciiBytes(srcBegin: Int, srcEnd: Int, dst: ByteArray, dstBegin: Int) {
+        (this as java.lang.String).getBytes(srcBegin, srcEnd, dst, dstBegin)
```

Comment:
Is this explicit cast necessary?

Reply:
Yes. `kotlin.String` does not expose this method (probably because it was deprecated as of JDK 1.1).